handsontable/src/plugins/copyPaste/copyPaste.js

```diff
@@ -338,35 +383,48 @@ export class CopyPaste extends BasePlugin {
    * Prepares copyable text from the cells selection in the invisible textarea.
    */
   setCopyableText() {
-    const selRange = this.hot.getSelectedRangeLast();
+    const selectionRange = this.hot.getSelectedRangeLast();
 
-    if (!selRange) {
+    if (!selectionRange) {
       return;
     }
 
-    const topStart = selRange.getTopStartCorner();
-    const bottomEnd = selRange.getBottomEndCorner();
-    const startRow = topStart.row;
-    const startCol = topStart.col;
-    const endRow = bottomEnd.row;
-    const endCol = bottomEnd.col;
-    const finalEndRow = Math.min(endRow, startRow + this.rowsLimit - 1);
-    const finalEndCol = Math.min(endCol, startCol + this.columnsLimit - 1);
+    this.#copyableRangesFactory.setSelectedRange(selectionRange);
 
-    this.copyableRanges.length = 0;
+    const groupedRanges = new Map([
+      ['headers', null],
+      ['cells', null],
+    ]);
 
-    this.copyableRanges.push({
-      startRow,
-      startCol,
-      endRow: finalEndRow,
-      endCol: finalEndCol
-    });
+    if (this.#copyMode === 'column-headers-only') {
+      groupedRanges.set('headers', this.#copyableRangesFactory.getMostBottomColumnHeadersRange());
+
+    } else {
+      if (this.#copyMode === 'with-column-headers') {
+        groupedRanges.set('headers', this.#copyableRangesFactory.getMostBottomColumnHeadersRange());
+
+      } else if (this.#copyMode === 'with-column-group-headers') {
+        groupedRanges.set('headers', this.#copyableRangesFactory.getAllColumnHeadersRange());
+      }
+
+      groupedRanges.set('cells', this.#copyableRangesFactory.getCellsRange());
+    }
+
+    this.copyableRanges = Array.from(groupedRanges.values())
+      .filter(range => range !== null)
+      .map(({ startRow, startCol, endRow, endCol }) => ({ startRow, startCol, endRow, endCol }));
 
     this.copyableRanges = this.hot.runHooks('modifyCopyableRange', this.copyableRanges);
 
-    if (endRow !== finalEndRow || endCol !== finalEndCol) {
-      this.hot
-        .runHooks('afterCopyLimit', endRow - startRow + 1, endCol - startCol + 1, this.rowsLimit, this.columnsLimit);
+    const cellsRange = groupedRanges.set('cells');
```

Comment:
Shouldn't this be
```suggestion
    const cellsRange = groupedRanges.get('cells');
```
?

Reply:
Thanks for spotting that! I've added missing tests for that logic (2e08acd1c298a181c74c3b27383028188ec41720).